### ISSUE TITLE
MASS: Masked Sequence to Sequence Pre-training for Language Generation

### ISSUE THREAD

**Issue by noowad93** [noowad93/ml-paper-summaries#1]
# 논문 정보 (URL, 저자 등)

[https://arxiv.org/abs/1905.02450](https://arxiv.org/abs/1905.02450)

Microsoft

# 개요

MASS는 encoder-decoder 프레임워크를 이용하여 랜덤하게 마스크된 문장의 fragment를 인코더에 입력했을 때 디코더에서 마스크된 fragment를 출력하도록 설계된 모델입니다. 

BERT나 GPT-2 등과 같은 인코더/디코더만 단독으로 학습된 모델과 달리 MASS는 2가지 이점을 갖고 있습니다. 

1. 인코더 측에서 마스크된 문장 fragment를 예측하게 함으로써, 디코더가 마스크된 토큰을 예측하게 하기 위해서 인코더가 마스크되지 않은 토큰들의 의미를 이해할 수 있도록 합니다.
2. 인코더 측에서 마스크되지 않은 토큰들을 디코더 측에서는 마스킹함으로써, 디코더가 다음 토큰을 예측할 때 source 문장에 더 많이 의존할 수 있게 되고, 인코더와 디코더간의 joint 학습이 더 용이해집니다.

 사전학습한 MASS를 zero/low resource 생성 태스크로 파인튜닝해본 결과, 기계 번역, 문서 요약, 답변 생성 등 다양한 태스크에서 사전 학습을 하지 않은 모델을 이용했을 때 보다 큰 성능 향상을 보였습니다.

# 모델

**모델 구조**

<img width="663" alt="스크린샷 2020-08-27 오후 4 14 22" src="https://user-images.githubusercontent.com/49383323/91409439-5f447000-e880-11ea-9c5d-188ab38b4dbf.png">

- 인코더에서 마스킹된 sentence fragment가 입력되고, 디코더에서는 반대로 소스 문장에서 마스킹 되지 않았던 문장에 마스킹을 하고, source sentence에서 마스킹된 부분을 예측합니다.
- 단, 인코더와 디코더 사이에 Bahdanau attention을 추가해서 토큰 사이의 alignment를 학습할 수 있도록 했고, encoder-decoder의 backbone 모델로는 Transformer를 사용합니다.

**사전 학습**

<img width="374" alt="스크린샷 2020-08-27 오후 4 14 54" src="https://user-images.githubusercontent.com/49383323/91409500-73886d00-e880-11ea-8b63-58de3652fa13.png">

- 위와 같이 sentence fragment이 마스킹된 문장을 조건부로 하여 마스킹된 fragment가 생성될 log-likelihood를 최대화하는 방식으로 사전학습을 진행합니다.

<img width="764" alt="스크린샷 2020-08-27 오후 4 15 23" src="https://user-images.githubusercontent.com/49383323/91409545-84d17980-e880-11ea-80c0-1b838357a299.png">

- 사실, BERT나 GPT 등은 MASS의 특별한 케이스라 할 수 있습니다.
    - sentence fragment의 길이 k를 1로 둔 경우는 BERT의 masked LM과 동일한 objective로 학습한다고 할수 있습니다.
        - 디코더가 있다고 해도, 실제로 decoder의 input tokens는 전부 마스킹 되어있으므로, decoder는 그 자체로 non-liear classifier라 볼 수 있고, BERT의 마지막 Linear Lyaer+ softmax와 동일한 기능을 한다고 볼 수 있습니다.
    - k를 전체 문장 길이 m으로 둔 경우는 GPT의 objective와 동일합니다.
        - encoder의 입력이 전부 마스킹 되어있으므로, 인코더로부터의 정보는 null information이기 때문에 사실상 GPT와 동일합니다.
    - 다만, 기존 방법들은 encoder와 decoder가 별도로 학습되기 때문에 생성 태스크에 적합하다고 할 수없습니다.

# 데이터

- WMT News Crawl datasets의 모든 monolingual data를 이용했습니다.
    - 영어, 프랑스어, 독일어 문장을 각각 190M, 62M, 270M 문장 포함하고 있습니다.
- 또한, low-resource 데이터에 대한 성능 검증을 위해, Romanian 언어도 사전학습에 포함을 했습니다.
    - 2.9M의 문장을 포함시켰습니다.
- 길이 175이상의 문장을 제거했고, BPE (Sennrich et al. 2016)을 이용한 60000개의 서브워드 기반으로 source, target 문장을 토크나이징하고, source와 target embedding을 joint하게 학습시켰습니다.
- 단, Text Summarization/Response Generation 태스크에서는 영어 문장만 이용했습니다.

# 실험

**모델 구조**

- 베이직 모델 설정으로는 각 6 레이어짜리 Transformer encoder, decoder를 이용했습니다.
    - embedding/hidden size로는 1024, feed-forward filter size로는 4096을 이용했습니다.
- machine translation task에서 source/target 언어를 구분하기 위해 문장의 각 토큰에 language embedding을 더하는 구조를 취했습니다.

**실험 세부사항**

- 특정 연속된 토큰을 마스킹하기 위해 다음과 같은 전략을 취했습니다.
    - 랜덤하게 마스킹할 시작 토큰 포지션 u를 고릅니다.
    - BERT와 동일하게, 80%는 [M] 토큰으로 치환, 10%는 랜덤 토큰으로 치환, 10%는 변경하지 않습니다.
- 계산 코스트를 줄이기 위해 다음과 같은 전략을 취했습니다.
    - 디코더에서 masked tokens를 실제로 학습에 사용하지 않고, positional embedding 값만 조정했습니다.
        - 이런 전략을 취해도 모델 성능은 거의 달라지지 않았습니다.
- Adam Optimizer with lr 1e-4를 이용해서 사전 학습을 진행했습니다.

**파인 튜닝 실험 결과**

- Unsupervised NMT, Text Summarization 등의 태스크에서 SOTA 성능을 기록했습니다.
- Response Generation 태스크에서 기존의 베이스라인들보다 더 낮은 perplexity를 기록했습니다.

<img width="402" alt="스크린샷 2020-08-27 오후 4 17 11" src="https://user-images.githubusercontent.com/49383323/91409714-c6622480-e880-11ea-9633-145d10fbab3d.png">

**Ablation Study**

- MASS는 consecutive tokens를 예측하기 때문에, 단순히 discrete tokens를 예측하는 태스크보다 language modeling의 capability가 좋습니다.
- source sentence 중 마스킹되지 않은 부분을 디코더에 다 입력하는 경우보다, 마스킹되지 않은 부분은 디코더에 입력하지 않는 경우가 더 성능이 좋습니다.

# 논의할만한 내용

- 마스킹된 문장을 입력했을 때 그 마스킹 부분에 들어갈만한 적절한 토큰을 찾는 문제로 사전학습을 진행하기 때문에 template-based generation에 응용가능한 면이 있음.
- 위에도 언급했듯이 인코더에 [M] 토큰만 입력하도록 하고, 파인 튜닝을 시키면 GPT와 동일하게 사용이 가능함.
- 다만, k=1일때 BERT와 완전히 동일하다고 할수 있는지는 의문인게, BERT는 사전 학습시 discrete한 [mask]토큰이 여러개 있기 때문.
- Generating Wikipedia By Summarizing Long Sequences (Liu et al, 2018)라는 논문을 제외하고는, Transformer Decoder가 Transformer Encoder-Decoder보다 성능이 우수하다는 논문을 아직 본적이 없음.
    - 위의 논문 역시 매우 긴 입력 문장+ Text Summarization이라는 특정 태스크에 한정된 이야기기도함.

# 관련된 논문

- XLM (Lample & Conneau, 2019)는 encoder와 decoder를 별도로 학습시키고, 파인튜닝시에 attention을 학습시키면서 jointly 학습한걸로 논문에는 쓰여있음.
